### PR TITLE
Add HTS engine support and voice

### LINKED ISSUE #76: HTS and vocoder alternatives
# Introduction (to be expanded)
- waveform synthesis techniques: there are many ways to synthesize a waveform: diphone (e.g. cmu_us_kal16 in mimic) cluster units, clustergen (eg cmu_us_slt, vid_us_ap),  HTS... Each technique has pros and cons.
- hts voices. 
  - hts Advantages: small footprint good quality.
  - hts issues: a bit more robotic than unit selection based engines, this can be improved with a better vocoder. Not easy to find free vocoder implementations.
  - There is a version of Flite with hts engine support [here](http://flite-hts-engine.sp.nitech.ac.jp)
- HTS support would be good for mimic due to the small voice footprint.
- Finding and integrating a good vocoder in mimic could make a difference in voice quality with small footprint.
- Some discussion was going on [here](https://github.com/MycroftAI/mimic/issues/61#issuecomment-239725458) #61. 

(I will copy a summary of the discussion here)

### PR DESCRIPTION
**Pull request by zeehio**
This PR closes #76 .

It provides HTS engine support to mimic and a new voice. I used the `flite+hts_engine` software as base, although I changed the integration of the systems to make HTS a first class synthesis technique in `mimic` (defining the `hts_synth` function, as done in clustergen voices).

As an example, you can build mimic and then try:

- `./mimic -voice slt -t "Hello world. I like pizza."`
- `./mimic -voice slt_hts -t "Hello world. I like pizza."`

Both voices are built with the same dataset (SLT ARCTIC), one with the clustergen technique and the other with HTS. A demo of hts voices is available [here](http://flite-hts-engine.sp.nitech.ac.jp/).

While it is true that the quality of the clustergen voice may be better than the hts voice in this case, the size of the clustergen voice is ~25MB while the size of the hts voice is <2MB. Smaller voices may be better for embedded applications.

HTS voices can be improved to make them sound more natural using a different vocoder (I am looking at [World](https://github.com/mmorise/World)), and that is something I would like to do in the future. Eventually, I would love to see a small-footprint version of mycroft voice that keeps the naturalness.

The PR is structured in several commits in order to make the review easier. If you see the "Changes" tab you can choose which commit to view. Reviewing them one by one may be easier than all of them at once.

Note that by default HTS support is compiled. If `libHTSEngine` is found on the system it is used, otherwise we use the bundled HTS engine available under `src/hts/hts_engine_API`

I have trained HTS voice models in the past, so I wanted to have HTS support in order to work on speech models in other languages.

### CODE REVIEW

**Comment by forslund** [comment]
Awesome! Will review during the week.

**Comment by forslund** [comment]
@zeehio I've finally starting to move on this.

So far only two minor notes:
As far as I can tell the voice file `cmu_us_slt_hts.htsvoice` can't be loaded like the other voice-files, only the precompiled version works. This is fine but maybe it should be mentioned in the readme?

In the rest of mimic the references to flite` has been changed to `mimic` (such as function names and file names). Should this be done for this submodule as well?

Now to the tricky part of the actual code =)

**Comment by forslund** [comment]
Looking further into the code I see how the file is loaded. I think it should be mentioned that it can't be loaded like the `.flitevox` files and that `-voicedump` can't be used with `slt_hts`.

**Comment by zeehio** [comment]
I will add to the documentation the things you mention about voice loading limitations. As you say the behaviour is different and needs to be documented.

About renaming, I don't mind that much and I imagine that if names are different it will be harder to merge this code with future flite versions, if we ever want to. But if for consistency you like better the renaming I can do it without issues :+1: .

**Comment by zeehio** [comment]
By the way, ignore the `hts_engine_API` subdirectory in the code review (unless you want to review external libraries). It is only compiled if the `libHTSEngine1` library is not found on the system.

**Comment by forslund** [comment]
Good to know. I get what you mean with the problem to merge back into the flite. Consistency is nice but not that important. This is very isolated so I don't think it's really a problem. Let's keep it the way it is.

File: lang/usenglish/us_ffeatures.c [hunk 1]
```diff
@@ -202,6 +246,261 @@ const cst_val *cg_content_words_in_phrase(const cst_item *p)
...
+  number of words in the sentence, or the total number of syllables. The
+  following functions extract those features for US English texts.
+ */
+
+int item_after_length(const cst_item *n)
+{
+    int i = 0;
+    if (n == NULL)
+        return 0;
+    for (; n; n = n->n, i++);
```
Comment (forslund) [inline lang/usenglish/us_ffeatures.c]: This is mainly my preference but I would prefer more verbosity in these for loops
```c
for (; n; n = n->n, i++)
{
    ; // do nothing
}
```

**Comment by forslund** [comment]
I've gone through this as thoroughly as possible and I've found no obvoius errors. Feel free to merge the branch after updating the documentation.

**Pull request merged**